Add support for year-long modules to module planner

jeffsieu opened the pull request:
## Context
Resolves #1912

## Implementation

### Before adding

![image](https://user-images.githubusercontent.com/8487294/96348312-6ba7b500-10da-11eb-9ef9-b478bf88fe8d.png)

### After adding

![image](https://user-images.githubusercontent.com/8487294/96347953-e4f1d880-10d7-11eb-889a-3e392a3c98cc.png)

Year long modules are represented as modules in semester 0.

### Adding year-long modules

Year-long modules are moved to the year-long module list whenever added from Semester 1 or 2. The year-long semester is hidden unless it is not empty, or being dragged over.

### Dragging modules

- Semester-long modules can only be dragged between regular semesters, exemptions and plan-to-take semesters.
- Year-long modules can only be dragged between year-long semesters, exemptions and plan-to-take semesters.

Upon dragging, their type (year-long or semester-long) is stored as part of their `draggableId`, which is then used to update `PlannerContainer.state.draggedModuleType`. This prompts children `PlannerYear` components to display their year-long semester containers as drag targets.


### Misc
- "+ Add modules" section removed to avoid clutter. Also avoids the case where non year-long modules are added from the year-long "add modules" section
- Adds `type` attribute to `Droppable` targets to disallow dragging of modules to the lists of the wrong module type.
- Some test cases are modified to due to the extra "semester".

jeffsieu committed:
Add support for year-long modules to module planner

Represents year long modules as modules in semester
0. Whenever new modules are added, they are checked
for whether they are year long modules. If they are,
they are moved to the year long "semester".

jeffsieu committed:
Fix module planner dragging

jeffsieu committed:
Fix dragging planner modules to bin

jeffsieu committed:
Add test cases for isYearLong and isYearLongModule

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

ZhangYiJiang commented on website/src/utils/planner.ts:
```diff
@@ -3,11 +3,14 @@ import { ModuleCode, PrereqTree, Semester } from 'types/modules';
 // outside the normal planner. "Exemption" modules are coded as -1 year so
 // they can always be used to fulfill prereqs, while "plan to take" modules use
 // 3000 so they can never fulfill prereqs
+export const YEAR_LONG_SEMESTER: Semester = 0;
```
This is problematic - 0 comes before all the other semesters, so a year long module would fulfill prerequisites for other modules taken in the same year, when the opposite should be true. In general we might want to re-think this hack of using specially coded semester values - it might be better to type this instead 

``` ts 
type Semester = 1 | 2 | 3 | 4;
type PlannerSemester = Semester | 'yearLong' | 'exemption' | 'planToTake';
``` 

but this will have other impact with regards to sorting

ZhangYiJiang replied:
Bump

ZhangYiJiang commented on website/src/views/planner/PlannerContainer/PlannerContainer.tsx:
```diff
@@ -100,26 +109,39 @@ export class PlannerContainerComponent extends React.PureComponent<Props, State>
-          this.props.fetchModule(moduleCode);
+          this.props.fetchModule(moduleCode).then((fetchedModule) => {
+            if (isYearLongModule(fetchedModule)) {
+              this.props.moveYearLongModule(year, semester, moduleCode);
```
If I understand the UI correctly, the way to add a year long module is to add it to either semester, then this code will move it to the year long section. 

That's a bit awkward - it means for at least one render the module will be in the wrong section, and there is also no real guarantee that the module will be added to the correct location, eg. if the module failed to load, which would leave the data in an inconsistent state since the module is still in the wrong location in the planner. 

We might want to instead think of adding `yearLong` flag to ModuleCondensed - we need this info for use in module search anyway, so it makes sense. This does require a change to the scraper though.

ZhangYiJiang reviewed (changes requested):
This looks great! I do think we might want to change the data structure from the scraper to allow us to do this more naturally. Let me know if you need help with that

jeffsieu commented on website/src/views/planner/PlannerYear.tsx:
```diff
@@ -78,7 +81,26 @@ export default class PlannerYear extends React.PureComponent<Props, State> {
+        <div
+          key={yearLongSemester}
+          className={classnames({
+            [styles.hideSemester]:
```
I tried that at first, but that causes the element to disappear right as the module is dropped on it. Then it will throw an error as the drop target is no longer there to receive it.

ZhangYiJiang replied:
Should add a comment then

jeffsieu committed:
Remove extraneous code

jeffsieu commented on website/src/utils/modules.ts:
```diff
@@ -138,3 +138,7 @@ export function getYearsBetween(minYear: string, maxYear: string): string[] {
   return Boolean(/[A-Z]+(5|6)\d{3}/i.test(module.moduleCode));
 }
+
+export function isYearLongModule(module: Module): boolean {
```
I suppose that could be done, but I was thinking of hiding it under a layer of abstraction.

jeffsieu commented on website/src/utils/planner.ts:
```diff
@@ -17,13 +20,19 @@ export const PLAN_TO_TAKE_SEMESTER = -2;
 // We assume iBLOCs takes place in special term 1
 export const IBLOCS_SEMESTER = 3;
 
+export const SEMESTER_LONG = 'SEMESTER';
```
This is used to set the type of PlannerModule as part of its id , so that the drop areas can only accept the currently dragged element if it should accept it (sem 1 shouldn't accept a year-long mod)

jeffsieu committed:
Merge branch 'planner-year-long-support' of https://github.com/jeffsieu/nusmods into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Include ModuleCondensed.yearLong in scraper

jeffsieu committed:
Merge branch 'planner-year-long-support' of https://github.com/jeffsieu/nusmods into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu commented:
@ZhangYiJiang How should I proceed with this PR? I really wish to see these changes be incorporated.

ZhangYiJiang commented on scrapers/nus-v2/src/tasks/CollateModules.ts:
```diff
@@ -118,10 +118,12 @@ const getModuleCondensed = ({
   moduleCode,
   title,
   semesters: semesterData.map((semester) => semester.semester),
+  yearLong: Boolean(attributes ? attributes.year : false),
```
`Boolean(undefined) === false` 

```suggestion
  yearLong: Boolean(attributes.year),
```

ZhangYiJiang commented on website/src/views/planner/PlannerModuleSelect.tsx:
```diff
@@ -121,7 +121,7 @@ export function PlannerModuleSelectComponent({
                       // Otherwise we use the input value - this allows the user to
                       // enter multiple
-                      onSelect(inputValue);
+                      onSelect(modules.find((module) => module.moduleCode === inputValue) ?? null);
```
This would break existing functionality where this field allows multiple module codes to be entered at once here https://github.com/nusmodifications/nusmods/blob/3e844019534912609e38cb546fa3e2b57f7279f6/website/src/views/planner/AddModule.tsx#L133

The logic for validating module code should be on the parent component, not the child.

`onSelect` did have incorrect typing - `ModuleCode` should just be `string`, sorry about that

jeffsieu replied:
Actually, how does the adding multiple module codes part work? Since `onSelect(inputValue)` seems to call this: https://github.com/nusmodifications/nusmods/blob/486cca0eda81a361485ee6e79ba219d67dd72c9e/website/src/views/planner/AddModule.tsx#L60-L69
which seems to add only 1 module

ZhangYiJiang replied:
It's in this function https://github.com/nusmodifications/nusmods/blob/4ba9c9bb893715c29c98300dff4b4032c650c62c/website/src/views/planner/PlannerContainer/PlannerContainer.tsx#L93. Again the types are wrong and the name is a bit misleading - moduleCode is matched to search for anything that looks like a module code

jeffsieu replied:
@ZhangYiJiang I changed the logic so that the module code extraction is done in `PlannerModuleSelect`, and `PlannerContainer.onAddModule` now only handles a single module at a time. i.e. `AddModuleData`'s type has also been modified. This decision was made because I needed the `yearLong` property of every module. And `PlannerModuleSelect` does have a reference to `modules` from which I can extract ModuleCondensed instances out of.

ZhangYiJiang commented on website/src/views/planner/PlannerSemester.tsx:
```diff
@@ -24,6 +30,7 @@ type Props = Readonly<{
 
   showModuleMeta?: boolean;
   className?: string;
+  draggedModuleType: string | null;
```
This should be explicitly typed as a union of all valid values. 

``` ts 
type ModuleTypes = 'YEAR_LONG' | 'SEMESTER_LONG'
``` 

It would also make it easier to follow if these are constructed and deconstructed in a type safe manner using functions 

``` ts 
function getDraggableId(moduleCode: ModuleCode, moduleType: ModuleType)

// Alternatively this can also return a tuple
function fromDraggableId(id: string): { moduleCode: ModuleCode, moduleType: ModuleType }
``` 

Also it seems that draggable type can be either `SEMESTER_LONG` or `YEAR_LONG` whereas droppable types are `SEMESTER_LONG`, `YEAR_LONG` and `DEFAULT`, but it's hard to tell what DEFAULT should mean - I think from reading the code it means it accepts all module types, in which case it should probably just be ALL - a comment would be really helpful

ZhangYiJiang reviewed (commented):
Sorry for not coming back to review this. This looks almost ready - just a few things 

- The reducers and actions that are now unused should be removed 
- Semester 0 should not be a valid global semester value, instead we can add something specifically for the planner 
- Drag and drop ID and module type should be strongly typed instead of just being strings

jeffsieu committed:
Add PlannerModuleSemester, ModuleType

PlannerModuleSemester represents a semester in
the module planner. PlannerSemesterModuleType represents
the module type accepted by a PlannerSemester
(semester or year long or both).

jeffsieu committed:
Update scrapers/nus-v2/src/tasks/CollateModules.ts

Co-authored-by: Zhang Yi Jiang <mediumdeviation@gmail.com>

jeffsieu committed:
Add workaround comment in PlannerYear

jeffsieu committed:
Merge branch 'planner-year-long-support' of https://github.com/jeffsieu/nusmods into planner-year-long-support

jeffsieu committed:
Module planner: Fix multiple module code entry

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

ZhangYiJiang commented on website/src/types/modules.ts:
```diff
@@ -50,7 +51,7 @@ export const WorkingDays: readonly Day[] = [
 export const DaysOfWeek: readonly Day[] = [...WorkingDays, 'Sunday'];
 
-export const Semesters: readonly Semester[] = [1, 2, 3, 4];
+export const Semesters: readonly Semester[] = [0, 1, 2, 3, 4];
```
Please revert this

ZhangYiJiang commented on website/src/selectors/planner.test.ts:
```diff
@@ -45,6 +45,7 @@ describe(getAcadYearModules, () => {
 
   test('should add semesters for empty years', () => {
     const emptyYear = {
+      [0]: [],
```
Please revert these changes

ZhangYiJiang commented on website/src/utils/planner.ts:
```diff
@@ -134,3 +163,15 @@ export function getTotalMC(
+  return module.moduleInfo ? Boolean(module.moduleInfo.attributes?.year) : false;
+}
+
+export function isSemester(plannerSemester: PlannerModuleSemester | string) { 
```
This makes the function a type guard which means you won't need to cast https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards 

```suggestion
export function isSemester(plannerSemester: PlannerModuleSemester): plannerSemester is Semester { 
```

ZhangYiJiang commented on website/src/selectors/planner.ts:
```diff
@@ -174,10 +173,10 @@ export const getIBLOCs = (state: State): PlannerModuleInfo[] => {
 
 export const getPlanToTake = (state: State): PlannerModuleInfo[] =>
-  mapNonTimetableModules(state, PLAN_TO_TAKE_YEAR, PLAN_TO_TAKE_SEMESTER);
+  mapNonTimetableModules(state, PLAN_TO_TAKE_YEAR, 'planToTake');
```
On L205 the code is still iterating over `Semesters` which does not include `yearLong`, so year long modules will never correctly count towards fulfilling prereqs.

ZhangYiJiang reviewed (commented):
Looking very close! Please revert some of the changes that are outdated, run the linter, and avoid using cast in favor of type guards, and this PR should be good to go

jeffsieu committed:
Merge branch 'master' into planner-year-long-support

jeffsieu committed:
Change isSemester to type guard

Co-authored-by: Zhang Yi Jiang <mediumdeviation@gmail.com>

jeffsieu committed:
Fix lint issues

jeffsieu committed:
Merge branch 'planner-year-long-support' of https://github.com/jeffsieu/nusmods into planner-year-long-support

jeffsieu commented:
@ZhangYiJiang Most issues should be resolved. There's currently still a circular dependency between types/planner.ts and types/reducers.ts.... I'm not sure how to resolve this without shifting the type declarations over from a file to the other.

yangshun commented:
Guys, merge this plox

ravern commented:
Hey, we're doing some clean ups over the summer. Closing this for now because both the issue and the PR are stale. Feel free to re-open this if you think otherwise!
